docs/enhancements/cluster_pool_hot_spares.md

```diff
@@ -0,0 +1,162 @@
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+    - [User Stories](#user-stories)
+  - [Proposal](#proposal)
+    - [`spec.runningCount`](#specrunningcount)
+    - [Decouple PowerState on Creation](#decouple-powerstate-on-creation)
+    - [ClusterClaim Assignment by `creationTimestamp`](#clusterclaim-assignment-by-creationtimestamp)
+    - [Hibernation Controller Must Account For `hibernateAfter` in Pool Clusters](#hibernation-controller-must-account-for-hibernateafter-in-pool-clusters)
+  - [Alternatives](#alternatives)
+    - [Don't Hibernate New ClusterDeployments](#dont-hibernate-new-clusterdeployments)
+    - [`runningPercent`](#runningpercent)
+
+## Summary
+
+Support configuring ClusterPools to maintain some number of clusters (`0 <= N <= spec.size`) in Running state.
+
+## Motivation
+
+Today ClusterPools immediately hibernate new clusters.
+This means a ClusterClaim is guaranteed to suffer at least Resuming time before the cluster is available for use.
+While this is still faster than having to create a cluster from scratch, it is a potentially-unnecessary delay.
+As an example, CI runs could be reduced by ~5m if a ClusterClaim were able to grab a Running cluster rather than having to Resume a Hibernating one.
+
+### Goals
+
+Ability to tune a ClusterPool such that a ClusterClaim gets a Running cluster immediately in the common case.
+
+### Non-Goals
+
+Solving world peace.
+
+
+### User Stories
+
+1. As a consumer of a (properly tuned) ClusterPool,
+
+   I want ClusterClaim to give me a Running cluster immediately in the common case,
+
+   So that I can get my work done more quickly.
+
+   (Ideally without sacrificing the cost savings of having some clusters in the pool Hibernating.)
+
+2. As a user of a (hypothetical future) ClusterDeployment (or sibling CRD) for a cluster type that doesn't support hibernation,
+
+   I want to be able to configure a ClusterPool that never hibernates its ClusterDeployments
+
+   So that I can use ClusterPools for that kind of ClusterDeployment.
+
+## Proposal
+
+### `spec.runningCount`
+
+Add a field to ClusterPool indicating a portion of the pool to maintain in Running state.
+`spec.runningCount` is an integer absolute number of Running clusters to strive for.
+It is constrained to `0 <= runningCount <= Size`.
+We could enforce this by webhook; or we could just update the spec accordingly.
+
+If unspecified, the default behavior is to hibernate all CDs in the pool, as today.
+This corresponds to `runningCount=0`, conveniently the golang nil value of an int.
+
+### Decouple PowerState on Creation
+
+Today we create all new ClusterDeployments with `powerState=Hibernating` and update that to `Running` at claim time.
+With this feature, we will create with `powerState` unset.
+Then in a separate code path:
+- Sort unassigned (installing + ready) CDs by `creationTimestamp`
```

Comment:
why do we include the installing cds in this? those can be acted upon when they are installed right?

Reply:
> Editing the power state of an installing cluster is cheap

This is based on the assumption that the hibernation controller does not act on the powerState request unless the cluster is installed.
i think a) we need to solidify that this behavior that powerState change is allowed on installing clusters b) hibernation controller only looks at changing state of installed clusters else no-ops

without those two, it seems like usage of an undefined behavior like the `partway hibernating`

Reply:
> assumption that the hibernation controller does not act on the powerState request unless the cluster is installed

General principle of idempotence / eventual consistency. If it doesn't behave nicely in these circumstances, it's a bug.

However, I'm liking the idea of reorganizing when we do what. I was talking to @suhanime this morning about [fixing broken claims (HIVE-1602)](https://issues.redhat.com/browse/HIVE-1602) and we discussed something like:

**Don't consider a CD assignable until it is Running.**

which, in the absence of `runningCount`, entails:

**For each unassigned claim, strive for an equal number of Running CDs.**

So we need to subdivide Installed, non-Deleting CDs by Hibernation status:
- **Running**: These can be assigned to unassigned claims.
- **Resuming**: These count against the number of unassigned claims, but aren't yet ready for assignment.
- **Hibernating**: These are eligible to be kicked over to `PowerState: Running` to (eventually) satisfy claims. We will do so when `numUnassignedClaims > (numResumingCDs + numRunningCDs)`

(Note that with this logic we can still starve claims if CDs get stuck Resuming -- but that's what HIVE-1602 is going to address.)

Factoring in `runningCount` simply means we strive for `numUnassignedClaims + runningCount` active CDs instead of just `numUnassignedClaims`.

Thoughts?

Reply:
> General principle of idempotence / eventual consistency. If it doesn't behave nicely in these circumstances, it's a bug.

Well technically there is a case to be made that such an Update should be rejected, and not eventual consistency (you can't hibernate something that hasn't reached the installed state in the first place). So unless this behavior is documented in the API, i don't think it's safe to make this assumption.